plugins/spi/SPIOutputTest.cpp

```diff
@@ -738,3 +751,388 @@ void SPIOutputTest::testCombinedAPA102Control() {
   // check if the output writes are 1
   OLA_ASSERT_EQ(1u, backend.Writes(1));
 }
+
+/**
+ * Test DMX writes in the individual APA102 Pixel Brightness mode.
+ */
+void SPIOutputTest::testIndividualAPA102ControlPixelBrightness() {
+  const uint16_t this_test_personality = SPIOutput::PERS_APA102PB_INDIVIDUAL;
+  // setup Backend
+  FakeSPIBackend backend(2);
+  SPIOutput::Options options(0, "Test SPI Device");
+  // setup pixel_count to 2 (enough to test all cases)
+  options.pixel_count = 2;
+  // setup SPIOutput
+  SPIOutput output(m_uid, &backend, options);
+  // set personality
+  output.SetPersonality(this_test_personality);
+
+  // simulate incoming DMX data with this buffer
+  DmxBuffer buffer;
+  // setup a pointer to the returned data (the fake SPI data stream)
+  unsigned int length = 0;
+  const uint8_t *data = NULL;
+
+  // test1
+  // setup some 'DMX' data
+  buffer.SetFromString("255, 1, 10, 100");
+  // simulate incoming data
+  output.WriteDMX(buffer);
+  // get fake SPI data stream
+  data = backend.GetData(0, &length);
+  // this is the expected spi data stream:
+  const uint8_t EXPECTED1[] = { 0, 0, 0, 0,               // StartFrame
+                                0xFF, 0x64, 0x0A, 0x01,   // first Pixel
+                                0x00, 0x00, 0x00, 0x00,   // second Pixel
+                                0};                       // EndFrame
+  // check for Equality
+  OLA_ASSERT_DATA_EQUALS(EXPECTED1, arraysize(EXPECTED1), data, length);
+  // check if the output writes are 1
+  OLA_ASSERT_EQ(1u, backend.Writes(0));
+
+  // test2
+  buffer.SetFromString("255, 255, 128, 0,  255, 10, 20, 30");
+  output.WriteDMX(buffer);
+  data = backend.GetData(0, &length);
+  const uint8_t EXPECTED2[] = { 0, 0, 0, 0,
+                                0xFF, 0x00, 0x80, 0xFF,
+                                0xFF, 0x1E, 0x14, 0x0A,
+                                0};
+  OLA_ASSERT_DATA_EQUALS(EXPECTED2, arraysize(EXPECTED2), data, length);
+  OLA_ASSERT_EQ(2u, backend.Writes(0));
+
+  // test3
+  // test what happens when only new data for the first leds is available.
+  // later data should be not modified so for pixel2 data set in test2 is valid
+  buffer.SetFromString("255, 34, 56, 78");
+  output.WriteDMX(buffer);
+  data = backend.GetData(0, &length);
+  const uint8_t EXPECTED3[] = { 0, 0, 0, 0,
+                                0xFF, 0x4E, 0x38, 0x22,
+                                0xFF, 0x1E, 0x14, 0x0A,
+                                0};
+  OLA_ASSERT_DATA_EQUALS(EXPECTED3, arraysize(EXPECTED3), data, length);
+  OLA_ASSERT_EQ(3u, backend.Writes(0));
+
+  // test4
+  // tests what happens if fewer then needed color information are received
+  buffer.SetFromString("7, 9");
```

Comment:
You ought to do this test with three channels, given you need four, and responding to three like other code would be a likely error.